akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala

```diff
@@ -177,87 +168,88 @@ private[akka] class RemoteWatcher(
   def quarantine(address: Address, uid: Option[Int]): Unit =
     remoteProvider.quarantine(address, uid)
 
-  def rewatchRemote(watchee: ActorRef, watcher: ActorRef): Unit =
-    if (watching.contains((watchee, watcher)))
-      watchRemote(watchee, watcher)
-    else
-      //has been unwatched inbetween, skip re-watch
-      log.debug("Ignoring re-watch after being unwatched in the meantime: [{} -> {}]", watcher.path, watchee.path)
+  def watchRemote(watchee: InternalActorRef, watcher: InternalActorRef): Unit = {
+    assert(watcher != self)
 
-  def watchRemote(watchee: ActorRef, watcher: ActorRef): Unit =
     if (watchee.path.uid == akka.actor.ActorCell.undefinedUid)
       logActorForDeprecationWarning(watchee)
-    else if (watcher != self) {
-      log.debug("Watching: [{} -> {}]", watcher.path, watchee.path)
-      addWatching(watchee, watcher)
 
-      // also watch from self, to be able to cleanup on termination of the watchee
-      context watch watchee
-      watching += ((watchee, self))
-    }
+    log.debug("Watching: [{} -> {}]", watcher.path, watchee.path)
+    watching.addBinding(watchee, watcher)
+    watchNode(watchee)
+
+    // add watch from self, this will actually send a Watch to the target when necessary
+    context watch watchee
+  }
 
-  def addWatching(watchee: ActorRef, watcher: ActorRef): Unit = {
-    watching += ((watchee, watcher))
+  def watchNode(watchee: InternalActorRef): Unit = {
     val watcheeAddress = watchee.path.address
-    if (!watchingNodes(watcheeAddress) && unreachable(watcheeAddress)) {
+    if (!watchByNodes.contains(watcheeAddress) && unreachable(watcheeAddress)) {
       // first watch to that node after a previous unreachable
       unreachable -= watcheeAddress
       failureDetector.remove(watcheeAddress)
     }
-    watchingNodes += watcheeAddress
+    watchByNodes.addBinding(watcheeAddress, watchee)
   }
 
-  def unwatchRemote(watchee: ActorRef, watcher: ActorRef): Unit =
+  def unwatchRemote(watchee: InternalActorRef, watcher: InternalActorRef): Unit = {
     if (watchee.path.uid == akka.actor.ActorCell.undefinedUid)
       logActorForDeprecationWarning(watchee)
-    else if (watcher != self) {
-      log.debug("Unwatching: [{} -> {}]", watcher.path, watchee.path)
-      watching -= ((watchee, watcher))
-
-      // clean up self watch when no more watchers of this watchee
-      if (watching.forall { case (wee, wer) ⇒ wee != watchee || wer == self }) {
-        log.debug("Cleanup self watch of [{}]", watchee.path)
-        context unwatch watchee
-        watching -= ((watchee, self))
-      }
-      checkLastUnwatchOfNode(watchee.path.address)
+    log.debug("Unwatching: [{} -> {}]", watcher.path, watchee.path)
+    watching.get(watchee) match {
+      case Some(watchers) if watchers == Set(watcher) ⇒
```

Comment:
This will be slow if the set is very large

I changed to `watchers -= watcher ; if (watchers.isEmpty)` 
This is the way it is done in Multimap, I think it is the fastest.